File: src/api/decoders/blueprint_change_proposal.ts
// Strict decoding for participant-only frozen Blueprint Change Proposal records.

import type { BlueprintChangeProposalAcceptanceRequest } from "../../../generated/api/BlueprintChangeProposalAcceptanceRequest";
import type { BlueprintChangeProposalAcceptedView } from "../../../generated/api/BlueprintChangeProposalAcceptedView";
import type { BlueprintChangeProposalComparisonView } from "../../../generated/api/BlueprintChangeProposalComparisonView";
import type { BlueprintChangeProposalCreateRequest } from "../../../generated/api/BlueprintChangeProposalCreateRequest";
import type { BlueprintChangeProposalDecisionView } from "../../../generated/api/BlueprintChangeProposalDecisionView";
import type { BlueprintChangeProposalDetailView } from "../../../generated/api/BlueprintChangeProposalDetailView";
import type { BlueprintChangeProposalPageView } from "../../../generated/api/BlueprintChangeProposalPageView";
import type { BlueprintChangeProposalSideView } from "../../../generated/api/BlueprintChangeProposalSideView";
import type { BlueprintChangeProposalSummaryView } from "../../../generated/api/BlueprintChangeProposalSummaryView";
import type { BlueprintForkApplySelection } from "../../../generated/api/BlueprintForkApplySelection";
import {
  DecodeError,
  decodeBoolean,
  decodeDictionary,
  decodeNullable,
  decodeRecord,
  decodeUuid,
} from "../decoder";
import {
  decodeCanonicalBlueprintCourse,
  decodeBlueprintComparisonView,
} from "./blueprint_comparison";
import { blueprintEditNumber, blueprintRevisionTuple, text } from "./blueprint_course";
import { decodeCourseClassification } from "./course_classification";
import { decodeCursorPage, decodeTimestamp, field, requireOnlyFields } from "./shared";

function names(value: unknown, path: string): { shortName: string; longName: string } {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["shortName", "longName"]);
  return {
    shortName: text(field(record, "shortName", path), `${path}.shortName`),
    longName: text(field(record, "longName", path), `${path}.longName`),
  };
}

function summary(value: unknown, path: string): BlueprintChangeProposalSummaryView {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "proposalId",
    "createdAt",
    "source",
    "sourceBlueprintEditNumber",
    "sourceNames",
    "target",
    "targetBlueprintEditNumber",
    "targetNames",
    "targetIsStale",
    "accepted",
  ]);
  const accepted = decodeNullable(
    field(record, "accepted", path),
    `${path}.accepted`,
    (item, itemPath) => {
      const row = decodeRecord(item, itemPath);
      requireOnlyFields(row, itemPath, ["acceptedAt", "target", "targetBlueprintEditNumber"]);
      return {
        acceptedAt: decodeTimestamp(field(row, "acceptedAt", itemPath), `${itemPath}.acceptedAt`),
        target: blueprintRevisionTuple(field(row, "target", itemPath), `${itemPath}.target`),
        targetBlueprintEditNumber: blueprintEditNumber(
          field(row, "targetBlueprintEditNumber", itemPath),
          `${itemPath}.targetBlueprintEditNumber`,
        ),
      };
    },
  );
  return {
    proposalId: decodeUuid(field(record, "proposalId", path), `${path}.proposalId`),
    createdAt: decodeTimestamp(field(record, "createdAt", path), `${path}.createdAt`),
    source: blueprintRevisionTuple(field(record, "source", path), `${path}.source`),
    sourceBlueprintEditNumber: blueprintEditNumber(
      field(record, "sourceBlueprintEditNumber", path),
      `${path}.sourceBlueprintEditNumber`,
    ),
    sourceNames: names(field(record, "sourceNames", path), `${path}.sourceNames`),
    target: blueprintRevisionTuple(field(record, "target", path), `${path}.target`),
    targetBlueprintEditNumber: blueprintEditNumber(
      field(record, "targetBlueprintEditNumber", path),
      `${path}.targetBlueprintEditNumber`,
    ),
    targetNames: names(field(record, "targetNames", path), `${path}.targetNames`),
    targetIsStale: decodeBoolean(field(record, "targetIsStale", path), `${path}.targetIsStale`),
    accepted,
  };
}

function selection(value: unknown, path: string): BlueprintForkApplySelection {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["sourceModuleLabels", "sourceAssessments", "layout"]);
  const copies = (
    name: "sourceModuleLabels" | "sourceAssessments",
    source: string,
    target: string,
  ): Array<Record<string, string | null>> => {
    const items = field(record, name, path);
    if (!Array.isArray(items)) throw new DecodeError(`${path}.${name}`, "an array");
    return items.map((item, index) => {
      const itemPath = `${path}.${name}[${index}]`;
      const row = decodeRecord(item, itemPath);
      requireOnlyFields(row, itemPath, [source, target]);
      return {
        [source]: decodeUuid(field(row, source, itemPath), `${itemPath}.${source}`),
        [target]: decodeNullable(field(row, target, itemPath), `${itemPath}.${target}`, decodeUuid),
      };
    });
  };
  const sourceModuleLabels = copies("sourceModuleLabels", "sourceModuleId", "targetModuleId");
  const sourceAssessments = copies("sourceAssessments", "sourceAssessmentId", "targetAssessmentId");
  const layout = decodeNullable(
    field(record, "layout", path),
    `${path}.layout`,
    (value, layoutPath) => {
      if (!Array.isArray(value) || value.length === 0)
        throw new DecodeError(layoutPath, "a nonempty complete layout");
      return value.map((rowValue, index) => {
        const rowPath = `${layoutPath}[${index}]`;
        const row = decodeRecord(rowValue, rowPath);
        requireOnlyFields(row, rowPath, ["module", "assessments"]);
        const destination = (
          input: unknown,
          destinationPath: string,
          module: boolean,
        ): Record<string, string> => {
          const destination = decodeRecord(input, destinationPath);
          const kind = field(destination, "kind", destinationPath);
          const key =
            kind === "existing"
              ? module
                ? "targetModuleId"
                : "targetAssessmentId"
              : kind === "newFromSource"
                ? module
                  ? "sourceModuleId"
                  : "sourceAssessmentId"
                : null;
          if (key === null) throw new DecodeError(destinationPath, "an explicit destination kind");
          requireOnlyFields(destination, destinationPath, ["kind", key]);
          return {
            kind: String(kind),
            [key]: decodeUuid(
              field(destination, key, destinationPath),
              `${destinationPath}.${key}`,
            ),
          };
        };
        const assessmentsValue = field(row, "assessments", rowPath);
        if (!Array.isArray(assessmentsValue) || assessmentsValue.length === 0)
          throw new DecodeError(`${rowPath}.assessments`, "a nonempty array");
        return {
          module: destination(field(row, "module", rowPath), `${rowPath}.module`, true),
          assessments: assessmentsValue.map((entry, entryIndex) =>
            destination(entry, `${rowPath}.assessments[${entryIndex}]`, false),
          ),
        };
      });
    },
  );
  return {
    sourceModuleLabels: sourceModuleLabels as BlueprintForkApplySelection["sourceModuleLabels"],
    sourceAssessments: sourceAssessments as BlueprintForkApplySelection["sourceAssessments"],
    layout: layout as BlueprintForkApplySelection["layout"],
  };
}

function decision(value: unknown, path: string): BlueprintChangeProposalDecisionView {
  const record = decodeRecord(value, path);
  const kind = field(record, "kind", path);
  if (kind === "entire") {
    requireOnlyFields(record, path, ["kind"]);
    return { kind };
  }
  if (kind !== "selected") throw new DecodeError(`${path}.kind`, "entire or selected");
  requireOnlyFields(record, path, [
    "kind",
    "selection",
    "sourceShortName",
    "sourceLongName",
    "sourceClassification",
  ]);
  return {
    kind,
    selection: selection(field(record, "selection", path), `${path}.selection`),
    sourceShortName: decodeBoolean(
      field(record, "sourceShortName", path),
      `${path}.sourceShortName`,
    ),
    sourceLongName: decodeBoolean(field(record, "sourceLongName", path), `${path}.sourceLongName`),
    sourceClassification: decodeBoolean(
      field(record, "sourceClassification", path),
      `${path}.sourceClassification`,
    ),
  };
}

function side(
  value: unknown,
  path: string,
): Omit<BlueprintChangeProposalSideView, "modules" | "assessments"> & {
  modules: unknown;
  assessments: unknown;
} {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "revision",
    "blueprintEditNumber",
    "names",
    "classification",
    "modules",
    "assessments",
  ]);
  return {
    revision: blueprintRevisionTuple(field(record, "revision", path), `${path}.revision`),
    blueprintEditNumber: blueprintEditNumber(
      field(record, "blueprintEditNumber", path),
      `${path}.blueprintEditNumber`,
    ),
    names: names(field(record, "names", path), `${path}.names`),
    classification: decodeCourseClassification(
      field(record, "classification", path),
      `${path}.classification`,
    ),
    modules: field(record, "modules", path),
    assessments: field(record, "assessments", path),
  };
}

function comparison(value: unknown, path: string): BlueprintChangeProposalComparisonView {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "source",
    "target",
    "assessmentRelationships",
    "sharedQuestionIds",
    "sourceOnlyQuestionIds",
    "targetOnlyQuestionIds",
  ]);
  const source = side(field(record, "source", path), `${path}.source`);
  const target = side(field(record, "target", path), `${path}.target`);
  const checked = decodeBlueprintComparisonView(
    {
      left: {
        currentRevision: source.revision,
        names: source.names,
        blueprintEditNumber: source.blueprintEditNumber,
        modules: source.modules,
        assessments: source.assessments,
      },
      right: {
        currentRevision: target.revision,
        names: target.names,
        blueprintEditNumber: target.blueprintEditNumber,
        modules: target.modules,
        assessments: target.assessments,
      },
      assessmentRelationships: field(record, "assessmentRelationships", path),
      sharedQuestionIds: field(record, "sharedQuestionIds", path),
      leftOnlyQuestionIds: field(record, "sourceOnlyQuestionIds", path),
      rightOnlyQuestionIds: field(record, "targetOnlyQuestionIds", path),
    },
    path,
  );
  return {
    source: { ...source, modules: checked.left.modules, assessments: checked.left.assessments },
    target: { ...target, modules: checked.right.modules, assessments: checked.right.assessments },
    assessmentRelationships: checked.assessmentRelationships,
    sharedQuestionIds: checked.sharedQuestionIds,
    sourceOnlyQuestionIds: checked.leftOnlyQuestionIds,
    targetOnlyQuestionIds: checked.rightOnlyQuestionIds,
  };
}

function accepted(value: unknown, path: string): BlueprintChangeProposalAcceptedView {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "acceptedAt",
    "target",
    "targetBlueprintEditNumber",
    "decision",
    "appliedSelection",
    "newModules",
    "newAssessments",
    "resultingJson",
  ]);
  return {
    acceptedAt: decodeTimestamp(field(record, "acceptedAt", path), `${path}.acceptedAt`),
    target: blueprintRevisionTuple(field(record, "target", path), `${path}.target`),
    targetBlueprintEditNumber: blueprintEditNumber(
      field(record, "targetBlueprintEditNumber", path),
      `${path}.targetBlueprintEditNumber`,
    ),
    decision: decision(field(record, "decision", path), `${path}.decision`),
    appliedSelection: selection(
      field(record, "appliedSelection", path),
      `${path}.appliedSelection`,
    ),
    newModules: decodeDictionary(
      field(record, "newModules", path),
      `${path}.newModules`,
      decodeUuid,
    ),
    newAssessments: decodeDictionary(
      field(record, "newAssessments", path),
      `${path}.newAssessments`,
      decodeUuid,
    ),
    resultingJson: decodeCanonicalBlueprintCourse(
      field(record, "resultingJson", path),
      `${path}.resultingJson`,
    ),
  };
}

export function decodeBlueprintChangeProposalCreateRequest(
  value: unknown,
  path = "request",
): BlueprintChangeProposalCreateRequest {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "source",
    "sourceBlueprintEditNumber",
    "target",
    "targetBlueprintEditNumber",
  ]);
  return {
    source: blueprintRevisionTuple(field(record, "source", path), `${path}.source`),
    sourceBlueprintEditNumber: blueprintEditNumber(
      field(record, "sourceBlueprintEditNumber", path),
      `${path}.sourceBlueprintEditNumber`,
    ),
    target: blueprintRevisionTuple(field(record, "target", path), `${path}.target`),
    targetBlueprintEditNumber: blueprintEditNumber(
      field(record, "targetBlueprintEditNumber", path),
      `${path}.targetBlueprintEditNumber`,
    ),
  };
}

export function decodeBlueprintChangeProposalAcceptanceRequest(
  value: unknown,
  path = "request",
): BlueprintChangeProposalAcceptanceRequest {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "expectedTarget",
    "expectedTargetBlueprintEditNumber",
    "decision",
  ]);
  return {
    expectedTarget: blueprintRevisionTuple(
      field(record, "expectedTarget", path),
      `${path}.expectedTarget`,
    ),
    expectedTargetBlueprintEditNumber: blueprintEditNumber(
      field(record, "expectedTargetBlueprintEditNumber", path),
      `${path}.expectedTargetBlueprintEditNumber`,
    ),
    decision: decision(field(record, "decision", path), `${path}.decision`),
  };
}

export function decodeBlueprintChangeProposalPageView(
  value: unknown,
  path = "response",
): BlueprintChangeProposalPageView {
  const page = decodeCursorPage(value, path, summary);
  return { items: [...page.items], nextCursor: page.nextCursor };
}

export function decodeBlueprintChangeProposalDetailView(
  value: unknown,
  path = "response",
): BlueprintChangeProposalDetailView {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["proposal", "canAccept", "comparison", "accepted"]);
  return {
    proposal: summary(field(record, "proposal", path), `${path}.proposal`),
    canAccept: decodeBoolean(field(record, "canAccept", path), `${path}.canAccept`),
    comparison: comparison(field(record, "comparison", path), `${path}.comparison`),
    accepted: decodeNullable(field(record, "accepted", path), `${path}.accepted`, accepted),
  };
}

export function decodeBlueprintChangeProposalAcceptedView(
  value: unknown,
  path = "response",
): BlueprintChangeProposalAcceptedView {
  return accepted(value, path);
}

File: crates/browser-api-contract/src/blueprint_change_proposal.rs
//! Participant-only frozen Change Proposal review and receiving-owner decisions.

use crate::blueprint_course::{
    BlueprintComparisonAssessment, BlueprintComparisonAssessmentRelationship,
    BlueprintComparisonModule, BlueprintComparisonNames,
};
use question_model::{BlueprintMetadataEtag, BlueprintRevisionReference, Timestamp};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Exact saved content and independently recorded metadata pins; never editor JSON.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BlueprintChangeProposalCreateRequest {
    pub source: BlueprintRevisionReference,
    pub source_metadata_etag: BlueprintMetadataEtag,
    pub target: BlueprintRevisionReference,
    pub target_metadata_etag: BlueprintMetadataEtag,
}

/// Whole-Assessment granularity, explicit layout, independent names and classification.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum BlueprintChangeProposalDecisionView {
    Entire,
    Selected {
        selection: question_model::blueprint_course::BlueprintForkApplySelection,
        source_short_name: bool,
        source_long_name: bool,
        source_classification: bool,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BlueprintChangeProposalAcceptanceRequest {
    pub expected_target: BlueprintRevisionReference,
    pub expected_target_metadata_etag: BlueprintMetadataEtag,
    pub decision: BlueprintChangeProposalDecisionView,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalAcceptedSummaryView {
    pub accepted_at: Timestamp,
    pub target: BlueprintRevisionReference,
    pub target_metadata_etag: BlueprintMetadataEtag,
}

/// Opaque UUID handle conveys no authorization, identity or global sequence.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalSummaryView {
    pub proposal_id: String,
    pub created_at: Timestamp,
    pub source: BlueprintRevisionReference,
    pub source_metadata_etag: BlueprintMetadataEtag,
    pub source_names: BlueprintComparisonNames,
    pub target: BlueprintRevisionReference,
    pub target_metadata_etag: BlueprintMetadataEtag,
    pub target_names: BlueprintComparisonNames,
    pub target_is_stale: bool,
    pub accepted: Option<BlueprintChangeProposalAcceptedSummaryView>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalPageView {
    pub items: Vec<BlueprintChangeProposalSummaryView>,
    pub next_cursor: Option<String>,
}

/// Original ID-bearing inventory, labelled with its exact pinned basis.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalSideView {
    pub revision: BlueprintRevisionReference,
    pub metadata_etag: BlueprintMetadataEtag,
    pub names: BlueprintComparisonNames,
    pub classification: question_model::CourseClassification,
    pub modules: Vec<BlueprintComparisonModule>,
    pub assessments: Vec<BlueprintComparisonAssessment>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalComparisonView {
    pub source: BlueprintChangeProposalSideView,
    pub target: BlueprintChangeProposalSideView,
    /// Existing relationship DTO: left means frozen source, right means frozen target.
    pub assessment_relationships: Vec<BlueprintComparisonAssessmentRelationship>,
    pub shared_question_ids: Vec<question_model::QuestionId>,
    pub source_only_question_ids: Vec<question_model::QuestionId>,
    pub target_only_question_ids: Vec<question_model::QuestionId>,
}

/// Committed immutable outcome, never a later ordinary target-head reload.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalAcceptedView {
    pub accepted_at: Timestamp,
    pub target: BlueprintRevisionReference,
    pub target_metadata_etag: BlueprintMetadataEtag,
    pub decision: BlueprintChangeProposalDecisionView,
    pub applied_selection: question_model::blueprint_course::BlueprintForkApplySelection,
    pub new_modules: BTreeMap<
        question_model::BlueprintModuleReference,
        question_model::BlueprintModuleReference,
    >,
    pub new_assessments: BTreeMap<
        question_model::BlueprintAssessmentReference,
        question_model::BlueprintAssessmentReference,
    >,
    pub resulting_json: question_model::CanonicalBlueprintCourse,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueprintChangeProposalDetailView {
    pub proposal: BlueprintChangeProposalSummaryView,
    /// Presentation only. SQL independently authorizes and guards final acceptance.
    pub can_accept: bool,
    pub comparison: BlueprintChangeProposalComparisonView,
    pub accepted: Option<BlueprintChangeProposalAcceptedView>,
}
